src/snowflake/connector/time_util.py

```diff
@@ -94,3 +152,82 @@ def get_timing_millis(self) -> int:
                 "Trying to get timing before TimerContextManager has finished"
             )
         return self._end - self._start
+
+
+class TimeoutBackoffCtx:
+    """Base context for handling timeouts and backoffs on retries"""
+
+    def __init__(
+        self,
+        max_retry_attempts: int | None = None,
+        timeout: int | None = None,
+        backoff_policy: BackoffPolicy | None = None,
+    ) -> None:
+        self._backoff_policy = backoff_policy
+
+        self._current_retry_count = 0
+        self._current_sleep_time = INITIAL_TIMEOUT_SLEEP_TIME
+
+        self._max_retry_attempts = max_retry_attempts
+        # in seconds
+        self._timeout = timeout
+
+        # in milliseconds
+        self._start_time_millis = None
+
+    @property
+    def timeout(self) -> int | None:
+        return self._timeout
+
+    @property
+    def current_retry_count(self) -> int:
+        return int(self._current_retry_count)
+
+    @property
+    def current_sleep_time(self) -> int:
+        return int(self._current_sleep_time)
+
+    def set_start_time(self) -> None:
+        self._start_time_millis = get_time_millis()
+
+    def remaining_time_millis(self, timeout: int | None) -> int | None:
+        if timeout is None or self._start_time_millis is None:
+            return None
```

Comment:
Maybe an Exception would be better here? I think it would be an issue if we misused this class

Reply:
Makes sense, made the edit.